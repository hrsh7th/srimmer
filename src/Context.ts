import produce from "immer";
import equals from "shallowequal";

/**
 * Select function.
 */
export type Select<State, T = any> = (state: State) => T;

/**
 * select's state.
 */
export type SelectState = {
  state: any;
  version: number;
};

/**
 * Context.
 */
export class Context<State> {
  /**
   * current state.
   */
  private state?: State;

  /**
   * listeners.
   */
  private changed = (_: State) => {};

  /**
   * Consumer's select props.
   */
  private selects: Map<Select<State>, SelectState> = new Map();

  /**
   * set state.
   */
  public setState = (state: State) => {
    this.state = state;
  };

  /**
   * get state.
   */
  public getState = () => {
    return this.state;
  };

  /**
   * update state.
   */
  public updateState = (update: (state: State) => void) => {
    this.state = produce(this.state, state => {
      update(state as State);
    });

    Array.from(this.selects.entries()).forEach(([select, state]) => {
      const next = select(this.state!);
      if (!equals(next, state.state)) {
        state.state = next;
        state.version++;
      }
    });

    this.changed(this.state!);
  };

  /**
   * listen state change.
   */
  public listen(changed: (state: State) => void) {
    this.changed = changed;
  }

  /**
   * register select function.
   */
  public register(select: Select<State>) {
    this.selects.set(select, {
      state: select(this.state!),
      version: 1
    });
  }

  /**
   * getSelectState.
   */
  public getSelectState(select: Select<State>) {
    return this.selects.get(select)!;
  }
}
