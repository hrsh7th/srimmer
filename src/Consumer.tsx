import React from "react";
import { Context, Select } from "./Context";

export type ConsumerProps<State, T = any> = {
  select: Select<State, T>;
  children: (selected: T) => React.ReactNode;
};

export type ConsumerState = {
  bit: number;
};

export function createConsumer<State>(
  Consumer: React.Consumer<State>,
  context: Context<State>
) {
  /**
   * SrimerConsumer.
   */
  return class SrimmerConsumer<T> extends React.Component<
    ConsumerProps<State, T>,
    ConsumerState
  > {
    /**
     * initialize bitmask.
     */
    public constructor(props: ConsumerProps<State>) {
      super(props);
      context.register(props.select);
    }

    /**
     * rednder.
     */
    public render() {
      const { select, children } = this.props;
      const bitState = context.getBitState(select);
      return (
        <Consumer unstable_observedBits={bitState.bit}>
          {state => children((bitState.state = select(state)))}
        </Consumer>
      );
    }
  };
}
