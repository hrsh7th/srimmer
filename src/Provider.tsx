import React from "react";
import { produce } from "immer";
import { Context, Update } from "./Context";

export type ProviderProps<State> = {
  state: State;
  children: React.ReactNode;
};

export type ProviderState<State> = {
  state: State;
};

export function createProvider<State>(
  Provider: React.Provider<State>,
  context: Context<State>
) {
  /**
   * SrimmerProvider.
   */
  return class SrimmerProvider extends React.Component<
    ProviderProps<State>,
    ProviderState<State>
  > {
    /**
     * props to state.
     */
    public constructor(props: ProviderProps<State>) {
      super(props);
      this.state = { state: props.state };
      context.getState = this.getState;
      context.updateState = this.updateState;
    }

    /**
     * render.
     */
    public render() {
      return (
        <Provider value={this.state.state}>{this.props.children}</Provider>
      );
    }

    /**
     * state getter.
     */
    public getState = () => {
      return this.state.state;
    };

    /**
     * state updator.
     */
    public updateState = (update: Update<State>) => {
      this.setState({
        state: produce<State>(this.state.state, function(state) {
          update(state as State);
        })
      });
    };
  };
}
