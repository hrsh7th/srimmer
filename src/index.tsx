import React from "react";
import { Patch, PatchListener } from "immer";
import { Context, Select } from "./Context";
import { createConsumer } from "./Consumer";
import { createProvider } from "./Provider";

export { Patch };

export type Select<
  T extends React.ComponentType<{ children: any }>
> = T extends React.ComponentType<{
  children: (selected: infer R) => React.ReactNode;
}>
  ? R
  : never;

export function define<State extends object>() {
  const context = new Context<State>();
  const { Provider, Consumer } = React.createContext<State>({} as any);
  const SrimmerProvider = createProvider<State>(Provider, context);
  const SrimmerConsumer = createConsumer<State>(Consumer, context);

  return {
    Provider: SrimmerProvider,
    Consumer: SrimmerConsumer,
    get: context.getState,
    set: context.setState,
    update: context.updateState,
    applyPatches: context.applyPatches,
    onPatches: (patchListener: PatchListener) => {
      context.patchListener = patchListener;
    },
    select: function select<T>(select: Select<State, T>) {
      return (props: { children: (s: T) => React.ReactNode }) => (
        <SrimmerConsumer<T> select={select}>{props.children}</SrimmerConsumer>
      );
    }
  };
}
