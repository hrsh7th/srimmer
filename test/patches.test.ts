import { State, createState, remove } from "./fixtures";
import { define, Patch } from "../src";

test("patches", () => {
  const { update, onPatches, set } = define<State>();
  set(createState());
  onPatches(patches => {
    expect(patches.length).toBe(1);
  });
  update(state => (state.tasks[0].name = "modified task1"));
});

test("patches with recursive update", () => {
  const { update, set, onPatches } = define<State>();
  set(createState());
  onPatches(patches => {
    expect(patches.length).toBe(2);
  });
  remove(1, update);
});

test("apply patches", () => {
  const { update, set, get, applyPatches, onPatches } = define<State>();
  set(createState());
  let patches = [] as Patch[];
  onPatches((_, inverse) => (patches = [...patches, ...inverse]));
  remove(1, update);
  expect(get()!).toMatchSnapshot();
  applyPatches(patches);
  expect(get()!).toMatchSnapshot();
});
