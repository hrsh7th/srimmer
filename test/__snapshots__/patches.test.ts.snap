// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apply patches 1`] = `
Object {
  "option": Object {
    "name": "option name",
  },
  "tasks": Array [
    Object {
      "id": 2,
      "name": "task2",
      "status": Object {
        "done": false,
      },
    },
  ],
}
`;

exports[`apply patches 2`] = `
Object {
  "option": Object {
    "name": "option name",
  },
  "tasks": Array [
    Object {
      "id": 1,
      "name": "task1",
      "status": Object {
        "done": true,
      },
    },
    Object {
      "id": 2,
      "name": "task2",
      "status": Object {
        "done": false,
      },
    },
    Object {
      "id": 3,
      "name": "task1-1",
      "parentId": 1,
      "status": Object {
        "done": false,
      },
    },
  ],
}
`;
